add rowsToScroll to scrollUp/Down w/ fallback to system default

Don-Vito opened the pull request:
## Summary of the Pull Request
- The number of lines to move upon scroll up scroll down can be defined
  in ScrollUp and ScrollDown commands (parameter is called
  "rowsToScroll").
- If the number are not provided, use the system default (the one we are
  using for mouse scrolls), rather than 1 line.

## Validation Steps Performed
* Manual testing
* Added custom bindings for scroll commands with different values,
  verified they and the default appear and behave as expected
* Checked that invalid values are not allowed

Closes #5078

zadjii-msft reviewed (approved):
Re-using `rowsToScroll` like that is really quite clever, I like it. This is again one of those things where it might be hard to test the interaction between the `TermControl` and the `TerminalPage`, so it's probably fine without a unit test.

\<PM hat\> Make sure to update the schema and docs as well\</PM hat\>, but I think this looks good to go!

Don-Vito commented:
> 
> 
> Re-using `rowsToScroll` like that is really quite clever, I like it. This is again one of those things where it might be hard to test the interaction between the `TermControl` and the `TerminalPage`, so it's probably fine without a unit test.
> 
> <PM hat> Make sure to update the schema and docs as well</PM hat>, but I think this looks good to go!

Will be done - wanted to hear the initial feedback!

Unrelated question - I see that linter github action is failing on the lines that were not modified. Is there a planned effort to lint the entire code base?

zadjii-msft commented:
>  I see that linter github action is failing on the lines that were not modified. Is there a planned effort to lint the entire code base?

Sorry about that. Looks like it's configured incorrectly, I'm hoping that we'll get it sorted out later today. You shouldn't have to change anything for this PR, the `clang-format` formatting should be good enough for the C++ files. Sorry for the delay 😕

Don-Vito commented:
> 
> 
> > I see that linter github action is failing on the lines that were not modified. Is there a planned effort to lint the entire code base?
> 
> Sorry about that. Looks like it's configured incorrectly, I'm hoping that we'll get it sorted out later today. You shouldn't have to change anything for this PR, the `clang-format` formatting should be good enough for the C++ files. Sorry for the delay 😕

@zadjii-msft - I am not sure it is misconfigured: it looks there are some white-space alignment issues in the code.
Another question is regarding the schema in docs. Is it linted usually? While working on the documentation I noticed that:
* for some reason "ctrl+shift+up" is not considered a valid binding by the regex in the schema
* CloseOtherTabsAction, CloseTabsAfterAction define the the type to be either integer or null, where I believe it should be "null" rather than null. At least my linter is angry about it.

zadjii-msft commented:
Well, the linter _shouldn't_ be running for C++ files. It actually doesn't know how to lint them, but then erroneously falls back to EDITORCONFIG as a language, and then immediately explodes. I've got a fix over in #7930.

The schema wasn't formally linted, though it might be now that #7637 has merged.
* I think there's actually an issue tracking updating the regex in the schema floating around somewhere... 
* I'm pretty sure that `null` is what we want here, _not_ the string `"null"`. What error is your linter giving you?

Don-Vito commented:
> 
> 
> Well, the linter _shouldn't_ be running for C++ files. It actually doesn't know how to lint them, but then erroneously falls back to EDITORCONFIG as a language, and then immediately explodes. I've got a fix over in #7930.
> 
> The schema wasn't formally linted, though it might be now that #7637 has merged.
> 
>     * I think there's actually an issue tracking updating the regex in the schema floating around somewhere...
> 
>     * I'm pretty sure that `null` is what we want here, _not_ the string `"null"`. What error is your linter giving you?

@zadjii-msft - thanks for the clarification!
Regarding the error I am getting:
![image](https://user-images.githubusercontent.com/4639110/96132202-ca1a3b00-0f02-11eb-9571-28d56ecc5639.png)

WSLUser commented:
The linter should pass now if you rebase on top of master. It runs every time a new commit is entered and with the updated changes, the ccp files should no longer show any errors. I'm not sure how resw should be treated but most of them passed anyways.

miniksa commented on src/cascadia/TerminalApp/TerminalPage.h:
```diff
@@ -28,6 +28,14 @@ namespace winrt::TerminalApp::implementation
 
+    enum ScrollDirection : int
+    {
+        // Hack alert, we are going to use numeric values to compute the actual scroll delta,
```
I appreciate the call out, but some days I feel like this entire product is just one giant hack alert.

Don-Vito replied:
I am new here - so checking the limits 😄 . I will fix this.

Don-Vito replied:
Removed the hack. Sorry.

DHowett replied:
HAH! This isn't a moment you need to apologize for. Michael is saying that we've got enough hacks that are bigger than this that we don't need to call this one out, and instead just do to it :smile:

miniksa reviewed (approved):
Great, thanks! Just get that rebase/merge relative to master in here so the linter chills out and we'll be good to go.

Don-Vito commented:
@miniksa -Addressed the comment and the linter is better, but now the CI is failing on something that I am not sure related to my changes:

```
StartGroup: OutputTests::WinPtyWrite#metadataSet1
TAEF: Data[method]: 1
TAEF: Data[selection]: true
Verify: SUCCEEDED(TestData::TryGetValue(L"method", method)): Get which function mode we should use
Verify: SUCCEEDED(TestData::TryGetValue(L"selection", selection)): Get whether we should use selection.
Error: Verify: IsTrue(ret && actual == strlen(buf)): ERROR: WriteFile returned 0: actual=0 LastError=6 (select)
 [File: D:\a\1\s\src\host\ft_host\API_OutputTests.cpp, Function: WinPtyDoWriteTest, Line: 1030]
EndGroup: OutputTests::WinPtyWrite#metadataSet1 [Failed]
```

Any ideas on how to proceed? :smile:

zadjii-msft commented:
@Don-Vito Sometimes the CI is just flaky. Looks like it's passed on re-run. 

I'll give Dustin the day to block if he has any last concerns, but I think this is good to go IMO.

zadjii-msft commented:
@msftbot merge this at 5pm

zadjii-msft commented:
@msftbot merge this in 14 hours

DHowett reviewed (changes requested):
I don't love that we're exposing TermControl's "RowsToScroll" as part of its public API, but there's definitely worse things in our codebase.

I just realized, writing this comment, that RowsToScroll has a magic value of `-1` (`WHEEL_PAGESCROLL`) that it can receive from the win32 system parameters API.

It might be better to remove the dependency on TermControl's rows and calculate it using the same mechanism as TC instead, so that we can keep track of the places we're using the system parameter and where we should have to care about WHEEL_PAGESCROLL.

Don-Vito commented:
> 
> 
> I don't love that we're exposing TermControl's "RowsToScroll" as part of its public API, but there's definitely worse things in our codebase.
> 
> I just realized, writing this comment, that RowsToScroll has a magic value of `-1` (`WHEEL_PAGESCROLL`) that it can receive from the win32 system parameters API.
> 
> It might be better to remove the dependency on TermControl's rows and calculate it using the same mechanism as TC instead, so that we can keep track of the places we're using the system parameter and where we should have to care about WHEEL_PAGESCROLL.

Thanks for the feedback! :smile: I will fix it. The original idea to use the same value to provide a more consistent behavior with mouse scroll speed.

Don-Vito commented:
@DHowett - when doing the change you suggested, I have few questions:

1. The value of RowsToScroll is recomputed today in TC. If I move to my own computation, I guess we will need some recomputation mechanism as well.
2. As long as I call the same API as TC, I will still need to handle the WHEEL_PAGESCROLL return value. So while it will probably be more explicit the fact that we taking this into account persists.

Don-Vito commented:
@DHowett - I am not sure I got everything correctly :smile:. In any case I reverted the change in the TermControl API and compute the default rows to scroll in TerminalPage itself using the same mechanics. To preserve some level of reloading this value I do it upon settings load (in TermControl it is updated upon getting focus as well).

Don-Vito commented:
@DHowett - anything I can do to unblock this one? I am ready to give some extra to ensure we have hacktoberfest tree planted 😄

DHowett reviewed (approved):
Love it. Thanks so much!
